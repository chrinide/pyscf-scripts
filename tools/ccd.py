#!/usr/bin/env python

import numpy
from pyscf import ao2mo
from pyscf import lib
from pyscf.lib import logger
einsum = lib.einsum

def kernel(cc, eris, t1=None, t2=None, max_cycle=80, tol=1e-8, tolnormt=1e-8,
           verbose=logger.INFO):
    if verbose is None:
        verbose = cc.verbose
    if isinstance(verbose, logger.Logger):
        log = verbose
    else:
        log = logger.Logger(cc.stdout, verbose)

    t1, t2 = cc.get_init_guess(eris)[1:]
    eold = 0
    eccsd = 0
    if cc.diis:
        adiis = lib.diis.DIIS(cc, cc.diis_file)
        adiis.space = cc.diis_space
    else:
        adiis = lambda t1,t2,*args: (t1,t2)

    conv = False
    for istep in range(max_cycle):
        foo, fov, fvv = make_inter_F(cc, t1, t2, eris)
        t2new = update_amp_t2(cc, t1, t2, eris, foo, fov, fvv)
        normt = numpy.linalg.norm(t2new-t2)
        t1, t2 = t1, t2new
        t1new = t2new = None
        if cc.diis:
            t1, t2 = cc.diis(t1, t2, istep, normt, eccsd-eold, adiis)
        eold, eccsd = eccsd, cc.energy(t1, t2, eris)
        log.info('istep = %d  E(CCD) = %.15g  dE = %.9g  norm(t2) = %.6g',
                 istep, eccsd, eccsd - eold, normt)
        if abs(eccsd-eold) < tol and normt < tolnormt:
            conv = True
            break
    return conv, eccsd, t1, t2


def make_inter_F(cc, t1, t2, eris):
    nocc, nvir = t1.shape
    fock = eris.fock

    tau = t2 
    theta = tau*2.0 - tau.transpose(2,1,0,3)

    foo = fock[:nocc,:nocc].copy()
    foo[range(nocc),range(nocc)] = 0
    g2 = 2.0*numpy.asarray(eris.ovoo)
    g2 -= numpy.asarray(eris.ovoo).transpose(2,1,0,3)
    foo += einsum('iakb,jakb->ij', eris.ovov, theta)
    g2 = None

    fov = fock[:nocc,nocc:]
    g2 = None

    fvv = fock[nocc:,nocc:].copy()
    fvv[range(nvir),range(nvir)] = 0
    g2 = 2.0*numpy.asarray(eris.ovvv)
    g2 -= numpy.asarray(eris.ovvv).transpose(0,3,2,1)
    fvv -= einsum('icja,icjb->ab', theta, eris.ovov)
    return foo, fov, fvv

def update_amp_t2(cc, t1, t2, eris, foo, fov, fvv):
    nocc, nvir = t1.shape
    fock = eris.fock
    t2new = numpy.copy(eris.ovov)

    ft_ij = foo 
    ft_ab = fvv 
    tmp1 = einsum('bc,iajc->iajb', ft_ab, t2)
    tmp2 = einsum('kj,kbia->iajb', ft_ij, t2)
    tw = tmp1 - tmp2
    tmp1 = tmp2 = None

    theta = t2*2.0 - t2.transpose(2,1,0,3)
    tau = theta

    wovOV = numpy.copy(eris.ovov)
    wovOV += 0.5*einsum('jakc,ibkc->ibja', tau, eris.ovov)
    wovOV -= 0.5*einsum('jakc,ickb->ibja', t2, eris.ovov)

    tau = 0.5*t2
    woVoV = -numpy.asarray(eris.oovv).transpose(1,2,0,3)
    woVoV += einsum('jcka,ickb->ibja', tau, eris.ovov)
    tw += einsum('iakc,kcjb->iajb', t2, woVoV)
    tw += einsum('icka,kcjb->ibja', t2, woVoV)
    tw += einsum('iakc,kcjb->iajb', theta, wovOV)
    tau = theta = None

    t2new += tw + tw.transpose(2,3,0,1)
    tmp1 = tw = None
    tau = t2 
    tmp1 = einsum('la,jaik->jlik', t1, eris.ovoo)
    tw = eris.oooo + tmp1 + tmp1.transpose(2,3,0,1)
    tw += einsum('iajb,kalb->ikjl', eris.ovov, tau)
    t2new += einsum('kilj,kalb->iajb', tw, tau)
    t2new += cc.add_wvVvV(t1, t2, eris)

    mo_e = fock.diagonal()
    eia = (mo_e[:nocc,None] - mo_e[None,nocc:]).reshape(-1)
    t2new /= (eia[:,None] + eia[None,:]).reshape(nocc,nvir,nocc,nvir)
    return t2new


def energy(cc, t1, t2, eris):
    tau = t2
    theta = tau*2.0 - tau.transpose(2,1,0,3)
    e = numpy.einsum('iajb,iajb', eris.ovov, theta)
    return e

def get_nocc(mycc):
    if mycc._nocc is not None:
        return mycc._nocc
    elif isinstance(mycc.frozen, (int, numpy.integer)):
        nocc = int(mycc.mo_occ.sum()) // 2 - mycc.frozen
        assert(nocc > 0)
        return nocc
    else:
        occ_idx = mycc.mo_occ > 0
        occ_idx[list(mycc.frozen)] = False
        return numpy.count_nonzero(occ_idx)

def get_nmo(mycc):
    if mycc._nmo is not None:
        return mycc._nmo
    elif isinstance(mycc.frozen, (int, numpy.integer)):
        return len(mycc.mo_occ) - mycc.frozen
    else:
        return len(mycc.mo_occ) - len(mycc.frozen)


class CCD(lib.StreamObject):
    def __init__(self, mf, frozen=0, mo_coeff=None, mo_occ=None):
        if mo_coeff  is None: mo_coeff  = mf.mo_coeff
        if mo_occ    is None: mo_occ    = mf.mo_occ

        self.mol = mf.mol
        self._scf = mf
        self.verbose = self.mol.verbose
        self.stdout = self.mol.stdout
        self.max_memory = mf.max_memory

        self.max_cycle = 50
        self.conv_tol = 1e-7
        self.conv_tol_normt = 1e-5
        self.diis_space = 6
        self.diis_file = None
        self.diis_start_cycle = 1
        self.diis_start_energy_diff = 1e-2

        self.frozen = frozen

##################################################
# don't modify the following attributes, they are not input options
        self.mo_coeff = mo_coeff
        self.mo_occ = mo_occ
        self.emp2 = None
        self.e_corr = None
        self.converged = False
        self._nocc = None
        self._nmo = None
        self.t1 = None
        self.t2 = None

        self._keys = set(self.__dict__.keys())

    def dump_flags(self):
        log = logger.Logger(self.stdout, self.verbose)
        log.info('')
        log.info('******** %s flags ********', self.__class__)
        nocc = self.nocc
        nvir = self.nmo - nocc
        log.info('CCD nocc = %d, nvir = %d', nocc, nvir)
        if self.frozen is not 0:
            log.info('frozen orbitals %s', str(self.frozen))
        log.info('max_cycle = %d', self.max_cycle)
        log.info('conv_tol = %g', self.conv_tol)
        log.info('conv_tol_normt = %s', self.conv_tol_normt)
        log.info('diis_space = %d', self.diis_space)
        log.info('diis_start_cycle = %d', self.diis_start_cycle)
        log.info('diis_start_energy_diff = %g', self.diis_start_energy_diff)
        log.info('max_memory %d MB (current use %d MB)',
                 self.max_memory, lib.current_memory()[0])
        return self

    @property
    def ecc(self):
       return self.e_corr

    @property
    def e_tot(self):
        return self.e_corr + self._scf.e_tot

    nocc = property(get_nocc)
    @nocc.setter
    def nocc(self, n):
        self._nocc = n

    nmo = property(get_nmo)
    @nmo.setter
    def nmo(self, n):
        self._nmo = n

    get_nocc = get_nocc
    get_nmo = get_nmo

    def get_init_guess(self, eris=None):
        if eris is None: eris = self.ao2mo(self.mo_coeff)
        return self.init_amps(eris)
    def init_amps(self, eris):
        nocc = self.nocc
        nvir = self.nmo - nocc
        #nvir = mo_e.size - nocc
        mo_e = eris.fock.diagonal()
        eia = mo_e[:nocc,None] - mo_e[None,nocc:]
        t2 = numpy.empty((nocc,nvir,nocc,nvir))
        self.emp2 = 0
        for i in range(nocc):
            gi = eris.ovov[i]
            t2[i] = gi/lib.direct_sum('a,jb->ajb', eia[i], eia)
            theta = gi*2 - gi.transpose(2,1,0)
            self.emp2 += numpy.einsum('ajb,ajb->', t2[i], theta)
        lib.logger.info(self, 'Init t2, MP2 energy = %.15g', self.emp2)
        t1 = eris.fock[:nocc,nocc:] / eia
        t1[:,:] = 0.0
        return self.emp2, t1, t2

    energy = energy

#   def kernel(self, t1=None, t2=None):
#       eris = self.ao2mo()
#       self.dump_flags()
#       self._conv, self.ecc, self.t1, self.t2 = \
#               kernel(self, eris, t1, t2, max_cycle=self.max_cycle,
#                      tol=self.conv_tol, tolnormt=self.conv_tol_normt,
#                      verbose=self.verbose)
#       return self.ecc, self.t1, self.t2

    def kernel(self, t1=None, t2=None, eris=None):
        return self.ccd(t1, t2, eris)
    def ccd(self, t1=None, t2=None, eris=None):
        if self.verbose >= logger.WARN:
            self.check_sanity()
        self.dump_flags()

        if eris is None:
            eris = self.ao2mo()
        self.converged, self.e_corr, self.t1, self.t2 = \
                kernel(self, eris, t1, t2, max_cycle=self.max_cycle,
                       tol=self.conv_tol, tolnormt=self.conv_tol_normt,
                       verbose=self.verbose)
        if self.converged:
            logger.info(self, 'CCD converged')
        else:
            logger.note(self, 'CCSD not converged')
        if self._scf.e_tot == 0:
            logger.note(self, 'E_corr = %.16g', self.e_corr)
        else:
            logger.note(self, 'E(CCD) = %.16g  E_corr = %.16g',
                        self.e_tot, self.e_corr)
        return self.e_corr, self.t1, self.t2

    def ao2mo(self):
        return _ERIS(self)

    def add_wvVvV(self, t1, t2, eris):
        tau = t2 
        return numpy.einsum('icjd,acbd->iajb', tau, eris.vvvv)

    def diis(self, t1, t2, istep, normt, de, adiis):
        if (istep > self.diis_start_cycle and
            abs(de) < self.diis_start_energy_diff):
            t1t2 = numpy.hstack((t1.ravel(),t2.ravel()))
            t1t2 = adiis.update(t1t2)
            t1 = t1t2[:t1.size].reshape(t1.shape)
            t2 = t1t2[t1.size:].reshape(t2.shape)
            logger.debug(self, 'DIIS for step %d', istep)
        return t1, t2

CC = CCD

class _ERIS:
    def __init__(self, cc, mo_coeff=None):
        if mo_coeff is None:
            self.mo_coeff = mo_coeff = _mo_without_core(cc, cc.mo_coeff)
        else:
            self.mo_coeff = mo_coeff = _mo_without_core(cc, mo_coeff)
# Note: Always recompute the fock matrix because cc._scf.mo_energy may not be
# the eigenvalue of Fock matrix (eg when level shift is used in the scf object
# and the scf does not converge, mo_energy has the contribution of level shift).
        dm = cc._scf.make_rdm1(cc.mo_coeff, cc.mo_occ)
        fockao = cc._scf.get_hcore() + cc._scf.get_veff(cc.mol, dm)
        self.fock = reduce(numpy.dot, (mo_coeff.T, fockao, mo_coeff))

        nocc = cc.nocc
        nmo = cc.nmo
        nvir = nmo - nocc

        eri1 = ao2mo.incore.full(cc._scf._eri, mo_coeff)
        eri1 = ao2mo.restore(1, eri1, nmo)
        self.oooo = eri1[:nocc,:nocc,:nocc,:nocc].copy()
        self.ooov = eri1[:nocc,:nocc,:nocc,nocc:].copy()
        self.ovoo = eri1[:nocc,nocc:,:nocc,:nocc].copy()
        self.oovv = eri1[:nocc,:nocc,nocc:,nocc:].copy()
        self.ovov = eri1[:nocc,nocc:,:nocc,nocc:].copy()
        self.ovvv = eri1[:nocc,nocc:,nocc:,nocc:].copy()
        self.vvvv = eri1[nocc:,nocc:,nocc:,nocc:].copy()

def get_moidx(cc):
    moidx = numpy.ones(cc.mo_occ.size, dtype=numpy.bool)
    if isinstance(cc.frozen, (int, numpy.integer)):
        moidx[:cc.frozen] = False
    elif len(cc.frozen) > 0:
        moidx[list(cc.frozen)] = False
    return moidx
def _mo_without_core(cc, mo):
    return mo[:,get_moidx(cc)]

if __name__ == '__main__':
    from pyscf import gto
    from pyscf import scf, cc

    mol = gto.Mole()
    mol.verbose = 4
    mol.atom = '''
    O
    H 1 1.1
    H 1 1.1 2 104
    '''
    mol.basis = '6-31g'
    mol.build()

    rhf = scf.RHF(mol)
    rhf.scf() 

    mcc = CCD(rhf)
    mcc.frozen = 0
    mcc.kernel()

    mycc = cc.CCSD(rhf)
    mycc.frozen = 0
    old_update_amps = mycc.update_amps
    def update_amps(t1, t2, eris):
        t1, t2 = old_update_amps(t1, t2, eris)
        return t1*0, t2
    mycc.update_amps = update_amps
    mycc.kernel()
    
    print('Ref CCD correlation energy', mycc.e_corr)

