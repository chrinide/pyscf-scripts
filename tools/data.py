#!/usr/bin/env python

import numpy

nlm = numpy.zeros((56,3), dtype=numpy.int32)

# p's
nlm[1,0] =  1 #px
nlm[2,1] =  1 #py
nlm[3,2] =  1 #pz

# d's
nlm[4,0] =  2  #xx
nlm[5,1] =  2  #yy
nlm[6,2] =  2  #zz
nlm[7,0] =  1  #xy
nlm[7,1] =  1
nlm[8,0] =  1  #xz
nlm[8,2] =  1
nlm[9,1] =  1  #yz
nlm[9,2] =  1

# f's
nlm[10,0] = 3 #xxx
nlm[11,1] = 3 #yyy
nlm[12,2] = 3 #zzz    
nlm[13,0] = 2 #xxy
nlm[13,1] = 1
nlm[14,0] = 2 #xxz
nlm[14,2] = 1
nlm[15,1] = 2 #yyz
nlm[15,2] = 1 
nlm[16,0] = 1 #xyy
nlm[16,1] = 2 
nlm[17,0] = 1 #xzz
nlm[17,2] = 2 
nlm[18,1] = 1 #yzz
nlm[18,2] = 2
nlm[19,0] = 1 #xyz
nlm[19,1] = 1 
nlm[19,2] = 1 

# g's
nlm[20,0] = 4 #xxxx
nlm[21,1] = 4 #yyyy
nlm[22,2] = 4 #zzzz
nlm[23,0] = 3 #xxxy
nlm[23,1] = 1
nlm[24,0] = 3 #xxxz
nlm[24,2] = 1 
nlm[25,0] = 1 #xyyy
nlm[25,1] = 3 
nlm[26,1] = 3 #yyyz
nlm[26,2] = 1 
nlm[27,0] = 1 #xzzz 
nlm[27,2] = 3 
nlm[28,1] = 1 #yzzz
nlm[28,2] = 3
nlm[29,0] = 2 #xxyy
nlm[29,1] = 2 
nlm[30,0] = 2 #xxzz
nlm[30,2] = 2
nlm[31,1] = 2 #yyzz 
nlm[31,2] = 2 
nlm[32,0] = 2 #xxyz 
nlm[32,1] = 1
nlm[32,2] = 1
nlm[33,0] = 1 #xyyz
nlm[33,1] = 2 
nlm[33,2] = 1
nlm[34,0] = 1 #xyzz
nlm[34,1] = 1
nlm[34,2] = 2

