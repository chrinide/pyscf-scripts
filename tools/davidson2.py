#!/usr/bin/env python

from __future__ import division
import math
import numpy as np
import time

''' Block Davidson, Joshua Goings (2013)

    Block Davidson method for finding the first few
        lowest eigenvalues of a large, diagonally dominant,
    sparse Hermitian matrix (e.g. Hamiltonian)
'''

n = 1200                                        # Dimension of matrix
tol = 1e-8                              # Convergence tolerance
mmax = n//2                             # Maximum number of iterations  

''' Create sparse, diagonally dominant matrix A with 
        diagonal containing 1,2,3,...n. The eigenvalues
    should be very close to these values. You can 
    change the sparsity. A smaller number for sparsity
    increases the diagonal dominance. Larger values
    (e.g. sparsity = 1) create a dense matrix
'''

sparsity = 0.0001
A = np.zeros((n,n))
for i in range(0,n):
    A[i,i] = i + 1 
A = A + sparsity*np.random.randn(n,n) 
A = (A.T + A)/2 


k = 8                                   # number of initial guess vectors 
eig = 4                                 # number of eignvalues to solve 
t = np.eye(n,k)                 # set of k unit vectors as guess
V = np.zeros((n,n))             # array of zeros to hold guess vec
I = np.eye(n)                   # identity matrix same dimen as A

# Begin block Davidson routine

start_davidson = time.time()

for m in xrange(k,mmax,k):
    if m <= k:
        for j in xrange(0,k):
            V[:,j] = t[:,j]/np.linalg.norm(t[:,j])
        theta_old = 1 
    elif m > k:
        theta_old = theta[:eig]
    V,R = np.linalg.qr(V)
    T = np.dot(V[:,:(m+1)].T,np.dot(A,V[:,:(m+1)]))
    THETA,S = np.linalg.eig(T)
    idx = THETA.argsort()
    theta = THETA[idx]
    s = S[:,idx]
    for j in xrange(0,k):
        w = np.dot((A - theta[j]*I),np.dot(V[:,:(m+1)],s[:,j])) 
#        q = np.dot(np.linalg.inv(theta[j]*I - np.diag(np.diag(A))),w)
        q = w/(theta[j]-A[j,j])
        V[:,(m+j+1)] = q
    norm = np.linalg.norm(theta[:eig] - theta_old)
    if norm < tol:
        break

end_davidson = time.time()

# End of block Davidson. Print results.

print "davidson = ", theta[:eig],";",\
    end_davidson - start_davidson, "seconds"

# Begin Numpy diagonalization of A

start_numpy = time.time()

E,Vec = np.linalg.eig(A)
E = np.sort(E)

end_numpy = time.time()

# End of Numpy diagonalization. Print results.

print "numpy = ", E[:eig],";",\
     end_numpy - start_numpy, "seconds" 

########################################################

# Build a fake sparse symmetric matrix 
n = 1600
print('Dimension of the matrix',n,'*',n)
sparsity = 0.001
A = np.zeros((n,n))
for i in range(0,n) : 
    A[i,i] = i-9
A = A + sparsity*np.random.randn(n,n)
A = (A.T + A)/2

tol = 1e-9             # Convergence tolerance
mmax = 20              # Maximum number of iterations

# Setup the subspace trial vectors
k = 4
print ('No. of start vectors:',k)
neig = 3
print ('No. of desired Eigenvalues:',neig)
t = np.eye(n,k) # initial trial vectors
v = np.zeros((n,n)) # holder for trial vectors as iterations progress
I = np.eye(n) # n*n identity matrix
ritz = np.zeros((n,n))
f = np.zeros((n,n))
#-------------------------------------------------------------------------------
# Begin iterations  
#-------------------------------------------------------------------------------
start = time.time()
iter = 0
for m in range(k,mmax,k):
    iter = iter + 1
    print ("Iteration no:", iter)
    if iter==1:  # for first iteration add normalized guess vectors to matrix v
        for l in range(m):
            v[:,l] = t[:,l]/(np.linalg.norm(t[:,l]))
    # Matrix-vector products, form the projected Hamiltonian in the subspace
    T = np.linalg.multi_dot([v[:,:m].T,A,v[:,:m]]) # selects fastest evaluation order
    w, vects = np.linalg.eig(T) # Diagonalize the subspace Hamiltonian
    jj = 0
    s = w.argsort()
    ss = w[s]
    #***************************************************************************
    # For each eigenvector of T build a Ritz vector, precondition it and check
    # if the norm is greater than a set threshold.
    #***************************************************************************
    for ii in range(m): #for each new eigenvector of T
        f = np.diag(1./ np.diag((np.diag(np.diag(A)) - w[ii]*I)))
#        print (f)
        ritz[:,ii] = np.dot(f,np.linalg.multi_dot([(A-w[ii]*I),v[:,:m],vects[:,ii]]))
        if np.linalg.norm(ritz[:,ii]) > 1e-7 :
            ritz[:,ii] = ritz[:,ii]/(np.linalg.norm(ritz[:,ii]))
            v[:,m+jj] = ritz[:,ii]
            jj = jj + 1
    q, r = np.linalg.qr(v[:,:m+jj-1])
    for kk in range(m+jj-1):
        v[:,kk] = q[:,kk]
    for ii in range(neig):
        print (ss[ii])
    if iter==1: 
        check_old = ss[:neig]
        check_new = 1
    elif iter==2:
        check_new = ss[:neig]
    else: 
        check_old = check_new
        check_new = ss[:neig]
    check = np.linalg.norm(check_new - check_old)
    if check < tol:
        print('Block Davidson converged at iteration no.:',iter)
        break
end = time.time()
print ('Block Davidson time:',end-start)
start = time.time()
eig, eigvecs = np.linalg.eig(A)
end = time.time() 
s = eig.argsort()
ss = eig[s]
print('Exact Diagonalization:')
for ii in range(neig):    
    print(ss[ii])
#print(ss[:neig])
print ('Exact Diagonalization time:',end-start)
     
