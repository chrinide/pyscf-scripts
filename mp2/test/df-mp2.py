#!/usr/bin/env python

import numpy
from pyscf import gto, scf, lib, ao2mo
from pyscf.tools import molden
einsum = lib.einsum

name = 'ch4_df'

mol = gto.Mole()
mol.basis = 'aug-cc-pvtz'
mol.atom = '''
C  0.0000  0.0000  0.0000
H  0.6276  0.6276  0.6276
H  0.6276 -0.6276 -0.6276
H -0.6276  0.6276 -0.6276
H -0.6276 -0.6276  0.6276
'''
mol.charge = 0
mol.spin = 0
mol.symmetry = 1
mol.verbose = 4
mol.build()

mf = scf.RHF(mol)
mf = scf.density_fit(mf)
mf.auxbasis = 'aug-cc-pvtz-jkfit'
ehf = mf.kernel()

ncore = 1
nao, nmo = mf.mo_coeff.shape
nocc = mol.nelectron//2 - ncore
nvir = nmo - nocc - ncore
mo_core = mf.mo_coeff[:,:ncore]
mo_occ = mf.mo_coeff[:,ncore:ncore+nocc]
mo_vir = mf.mo_coeff[:,ncore+nocc:]
co = mo_occ
cv = mo_vir
eo = mf.mo_energy[ncore:ncore+nocc]
ev = mf.mo_energy[ncore+nocc:]
lib.logger.info(mf,"* Core orbitals: %d" % ncore)
lib.logger.info(mf,"* Virtual orbitals: %d" % (len(ev)))

naux = mf._cderi.shape[0]
dferi = numpy.empty((naux,nao,nao))
for i in range(naux):
    dferi[i] = lib.unpack_tril(mf._cderi[i])
eri_mo = einsum('rj,Qrs->Qjs', co, dferi)
eri_mo = einsum('sb,Qjs->Qjb', cv, eri_mo)

vv_denom = -ev.reshape(-1,1)-ev
t2 = numpy.zeros((nocc,nvir,nocc,nvir))
for i in range(nocc):
    eps_i = eo[i]
    i_Qv = eri_mo[:, i, :].copy()
    for j in range(nocc):
        eps_j = eo[j]
        j_Qv = eri_mo[:, j, :].copy()
        viajb = einsum('Qa,Qb->ab', i_Qv, j_Qv)
        vibja = einsum('Qb,Qa->ab', i_Qv, j_Qv)
        div = 1.0/(eps_i + eps_j + vv_denom)
        t2[i,:,j,:] += 2.0*einsum('ab,ab->ab', viajb, div) 
        t2[i,:,j,:] -= 1.0*einsum('ab,ab->ab', vibja, div) 

e_mp2 = einsum('iajb,Qia->Qjb', t2, eri_mo)
e_mp2 = numpy.einsum('Qjb,Qjb->', e_mp2, eri_mo, optimize=True)
lib.logger.info(mf,"!*** E(MP2): %12.8f" % e_mp2)
lib.logger.info(mf,"!**** E(HF+MP2): %12.8f" % (e_mp2+ehf))

den_file = name + '.den'
fspt = open(den_file,'w')
fspt.write('MP2\n')
fspt.write('1-RDM:\n')
occup = 2.0
norb = mol.nelectron//2
for i in range(norb):
    fspt.write('%i %i %.10f\n' % ((i+1), (i+1), occup))
fspt.write('t2_iajb:\n')
for i in range(nocc):
    for j in range(nvir):
        for k in range(nocc):
            for l in range(nvir):
                if (abs(t2[i,j,k,l]) > 1e-8):
                        fspt.write('%i %i %i %i %.10f\n' % ((i+1+ncore), \
                        (j+1+nocc+ncore), (k+1+ncore), (l+1+nocc+ncore), \
                        t2[i,j,k,l]*2.0))
fspt.close()                    
    
with open(name+'.mol', 'w') as f2:
    molden.header(mol, f2)
    molden.orbital_coeff(mol, f2, mf.mo_coeff, occ=mf.mo_occ)

