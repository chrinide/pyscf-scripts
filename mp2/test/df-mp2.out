#INFO: **** input file is /home/jluis/src/git/scripts/mp2/test/df-mp2.py ****
#!/usr/bin/env python

import numpy
from pyscf import gto, scf, lib, ao2mo
from pyscf.tools import molden
einsum = lib.einsum

name = 'ch4_df'

mol = gto.Mole()
mol.basis = 'aug-cc-pvtz'
mol.atom = '''
C  0.0000  0.0000  0.0000
H  0.6276  0.6276  0.6276
H  0.6276 -0.6276 -0.6276
H -0.6276  0.6276 -0.6276
H -0.6276 -0.6276  0.6276
'''
mol.charge = 0
mol.spin = 0
mol.symmetry = 1
mol.verbose = 4
mol.build()

mf = scf.RHF(mol)
mf = scf.density_fit(mf)
mf.auxbasis = 'aug-cc-pvtz-jkfit'
ehf = mf.kernel()

ncore = 1
nao, nmo = mf.mo_coeff.shape
nocc = mol.nelectron//2 - ncore
nvir = nmo - nocc - ncore
mo_core = mf.mo_coeff[:,:ncore]
mo_occ = mf.mo_coeff[:,ncore:ncore+nocc]
mo_vir = mf.mo_coeff[:,ncore+nocc:]
co = mo_occ
cv = mo_vir
eo = mf.mo_energy[ncore:ncore+nocc]
ev = mf.mo_energy[ncore+nocc:]
lib.logger.info(mf,"* Core orbitals: %d" % ncore)
lib.logger.info(mf,"* Virtual orbitals: %d" % (len(ev)))

naux = mf._cderi.shape[0]
dferi = numpy.empty((naux,nao,nao))
for i in range(naux):
    dferi[i] = lib.unpack_tril(mf._cderi[i])
eri_mo = einsum('rj,Qrs->Qjs', co, dferi)
eri_mo = einsum('sb,Qjs->Qjb', cv, eri_mo)

vv_denom = -ev.reshape(-1,1)-ev
t2 = numpy.zeros((nocc,nvir,nocc,nvir))
for i in range(nocc):
    eps_i = eo[i]
    i_Qv = eri_mo[:, i, :].copy()
    for j in range(nocc):
        eps_j = eo[j]
        j_Qv = eri_mo[:, j, :].copy()
        viajb = einsum('Qa,Qb->ab', i_Qv, j_Qv)
        vibja = einsum('Qb,Qa->ab', i_Qv, j_Qv)
        div = 1.0/(eps_i + eps_j + vv_denom)
        t2[i,:,j,:] += 2.0*einsum('ab,ab->ab', viajb, div) 
        t2[i,:,j,:] -= 1.0*einsum('ab,ab->ab', vibja, div) 

e_mp2 = einsum('iajb,Qia->Qjb', t2, eri_mo)
e_mp2 = numpy.einsum('Qjb,Qjb->', e_mp2, eri_mo, optimize=True)
lib.logger.info(mf,"!*** E(MP2): %12.8f" % e_mp2)
lib.logger.info(mf,"!**** E(HF+MP2): %12.8f" % (e_mp2+ehf))

den_file = name + '.den'
fspt = open(den_file,'w')
fspt.write('MP2\n')
fspt.write('1-RDM:\n')
occup = 2.0
norb = mol.nelectron//2
for i in range(norb):
    fspt.write('%i %i %.10f\n' % ((i+1), (i+1), occup))
fspt.write('t2_iajb:\n')
for i in range(nocc):
    for j in range(nvir):
        for k in range(nocc):
            for l in range(nvir):
                if (abs(t2[i,j,k,l]) > 1e-8):
                        fspt.write('%i %i %i %i %.10f\n' % ((i+1+ncore), \
                        (j+1+nocc+ncore), (k+1+ncore), (l+1+nocc+ncore), \
                        t2[i,j,k,l]*2.0))
fspt.close()                    
    
with open(name+'.mol', 'w') as f2:
    molden.header(mol, f2)
    molden.orbital_coeff(mol, f2, mf.mo_coeff, occ=mf.mo_occ)

#INFO: ******************** input file end ********************


System: ('Linux', 'silicio', '3.16.0-4-amd64', '#1 SMP Debian 3.16.51-3 (2017-12-13)', 'x86_64', '')  Threads 12
Python 2.7.15 (default, Oct 25 2018, 01:35:55) 
[GCC 8.2.0]
numpy 1.15.3  scipy 1.1.0
Date: Fri Dec  7 19:07:56 2018
PySCF version 1.6b
PySCF path  /home/jluis/src/pyscf/dev/pyscf
GIT ORIG_HEAD 45f68f7c1864ba6705547042f79d02f17d6d91ee
GIT HEAD      ref: refs/heads/dev
GIT dev branch  45f68f7c1864ba6705547042f79d02f17d6d91ee

[ENV] PYSCF_TMPDIR /scratch-ssd/jluis
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 5
[INPUT] num. electrons = 10
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry 1 subgroup None
[INPUT] Mole.unit = angstrom
[INPUT]  1 C      0.000000000000   0.000000000000   0.000000000000 AA    0.000000000000   0.000000000000   0.000000000000 Bohr
[INPUT]  2 H      0.627600000000   0.627600000000   0.627600000000 AA    1.185992115777   1.185992115777   1.185992115777 Bohr
[INPUT]  3 H      0.627600000000  -0.627600000000  -0.627600000000 AA    1.185992115777  -1.185992115777  -1.185992115777 Bohr
[INPUT]  4 H     -0.627600000000   0.627600000000  -0.627600000000 AA   -1.185992115777   1.185992115777  -1.185992115777 Bohr
[INPUT]  5 H     -0.627600000000  -0.627600000000   0.627600000000 AA   -1.185992115777  -1.185992115777   1.185992115777 Bohr

nuclear repulsion = 13.4720345873821
point group symmetry = Td, use subgroup D2
num. orbitals of irrep A = 36
num. orbitals of irrep B1 = 34
num. orbitals of irrep B2 = 34
num. orbitals of irrep B3 = 34
number of shells = 49
number of NR pGTOs = 158
number of NR cGTOs = 138
basis = aug-cc-pvtz
ecp = {}
CPU time:         0.23


******** <class 'pyscf.df.df_jk.DFHF'> ********
method = DFHF-SymAdaptedRHF-RHF
initial guess = minao
damping factor = 0
level shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
DIIS start cycle = 1
DIIS space = 8
SCF tol = 1e-09
SCF gradient tol = None
max. SCF cycles = 50
direct_scf = False
chkfile to save SCF result = /scratch-ssd/jluis/tmp_feu8I
max_memory 60000 MB (current use 67 MB)
Freeze 0 electrons in irreps []
    10 free electrons in irreps A B1 B2 B3
Set gradient conv threshold to 3.16228e-05
******** <class 'pyscf.df.df.DF'> ********
auxbasis = None
max_memory = 60000
_cderi_to_save = /scratch-ssd/jluis/tmpxy_fAo
init E= -40.0610425387277
HOMO (B1) = -0.501966018992381  LUMO (A) = 0.0218455901010336
cycle= 1 E= -40.2045950196473  delta_E= -0.144  |g|= 0.177  |ddm|= 1.35
HOMO (B1) = -0.534830430827343  LUMO (A) = 0.0307027185651426
cycle= 2 E= -40.212788926552  delta_E= -0.00819  |g|= 0.0495  |ddm|= 0.163
HOMO (B3) = -0.552121441015281  LUMO (A) = 0.0305730365064985
cycle= 3 E= -40.2135286043008  delta_E= -0.00074  |g|= 0.0187  |ddm|= 0.0409
HOMO (B1) = -0.545013840946933  LUMO (A) = 0.0308395506936041
cycle= 4 E= -40.2136362402862  delta_E= -0.000108  |g|= 0.00177  |ddm|= 0.014
HOMO (B3) = -0.545206619496253  LUMO (A) = 0.0308462202592942
cycle= 5 E= -40.2136382611159  delta_E= -2.02e-06  |g|= 0.000251  |ddm|= 0.00384
HOMO (B2) = -0.545262183184979  LUMO (A) = 0.0308438723722151
cycle= 6 E= -40.2136382884668  delta_E= -2.74e-08  |g|= 1.7e-05  |ddm|= 0.000499
HOMO (B1) = -0.545263779074789  LUMO (A) = 0.0308434719327027
cycle= 7 E= -40.2136382885713  delta_E= -1.05e-10  |g|= 9.51e-07  |ddm|= 4.47e-05
HOMO (B1) = -0.545263978941499  LUMO (A) = 0.0308434674051517
Extra cycle  E= -40.2136382885716  delta_E= -2.84e-13  |g|= 1.92e-07  |ddm|= 7.7e-06
converged SCF energy = -40.2136382885716
* Core orbitals: 1
* Virtual orbitals: 133
!*** E(MP2):  -0.20075398
!**** E(HF+MP2): -40.41439227
